### PR TITLE
Add support for (TARGET/SOURCE/TARGETS/SOURCES/CHANGED_TARGETS/CHANGED_SOURCES).relpath property

### LINKED ISSUE #396: relative path in build option
This issue was originally created at: 2003-01-21 05:53:00.
This issue was reported by: `nbecker`.
nbecker said at 2003-01-21 05:53:00
>> Repository ('../../src/fixed') 
>> Repository ('../../src/float') 
>> env = Environment(LIBS = 'libiTest', LIBPATH = ['.'], 
>CXXFLAGS = '-g') 
>> env.Library(target = 'libiTest', source = ['polint.cc', 
>'QPSKConstellation.cc', 'ODemod.cc', 'SerialFFTFilt.cc', 
>'FFTEqualizer.cc', 'iDFTEqualizer.cc', 'TWT.cc', 'FastTWT.cc']) 
>>  
>> ALLTESTS = [ 'Test32', 'Test33' ] 
>> for test in ALLTESTS: 
>> 	env.Program(test, test+'.cc') 
>        . 
>        . 
>        . 
>> Only one problem.  I don't want the argument to c++ to 
>contain the 
>> complete file path.  If it did, the executable would be harder 
>to 
>> debug if moved to a different location.  I'd want it to say: 
>>  
>> c++ -g -c -o polint.o ../../src/float/polint.cc 
>>  
>> Any ideas? 
> 
>Unfortunately, that's just the way it works right now.  We need 
>a way to 
>specify this, probably through a '.relpath' attribute that could be 
>used 
>for $TARGET(S) and $SOURCE(S) expansion: 
> 
>        $CC -c -o ${TARGET.relpath} ${SOURCES.relpath} 
> 
>Could you open up a feature request for this?  Thanks!

issues@scons said at 2003-01-21 05:53:00
>Converted from SourceForge tracker item 671788

garyo said at 2008-03-17 18:16:32
>SHould be feasible; try for 1.x.

garyo said at 2008-03-17 18:18:07
>p3.

garyo said at 2008-03-17 18:25:39
>2.x since this would be nice but it's not a must-have feature.

gregnoel said at 2008-12-26 13:20:19
>Adjust triage of issues.

Votes for this issue: 1.

### PR DESCRIPTION
**Pull request by DeeeeLAN**
Fixes #396 

Adds .relpath to all of TARGET/SOURCE/TARGETS/SOURCES/CHANGED_TARGETS/CHANGED_SOURCES

## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `CHANGES.txt` (and read the `README.rst`)
* [x] I have updated the appropriate documentation

### CODE REVIEW

File: CHANGES.txt [hunk 1]
```diff
@@ -35,6 +35,8 @@ RELEASE  VERSION/DATE TO BE FILLED IN LATER
     - Drop remaining definitions of dict-like has_key methods, since
       Python 3 doesn't have a dictionary has_key (maintenance)
 
+  From Dillan Mills:
+    - Add support for the TARGET.relpath property
```
Comment (bdbaddog) [inline CHANGES.txt]: Should actually be `TARGET/SOURCE/TARGETS/SOURCES/CHANGED_TARGETS/CHANGED_SOURCES`

**Comment by bdbaddog** [comment]
I've pushed some minor edits and a bit of a reorg.
Can you annotate the test's SConstruct as to what you're trying to do a bit?
I'm not sure I understand?
Target has two args is one supposed to be a file and one a directory?
Is the source supposed to be a directory?

**Comment by DeeeeLAN** [comment]
I originally copied the code from one of the Dir tests. Currently I am just using the three as a way to test a higher, lower, and current directories. We could rearrange them into whatever combination we wanted, and probably want to test both directories and files for targets and sources.

**Comment by bdbaddog** [comment]
> I originally copied the code from one of the Dir tests. Currently I am just using the three as a way to test a higher, lower, and current directories. We could rearrange them into whatever combination we wanted, and probably want to test both directories and files for targets and sources.

In theory you shouldn't need to specify the factory or the scanners..

**Comment by DeeeeLAN** [comment]
Yeah probably, although I tried once without the factory or scanners (outside the test, just running the SConstruct file manually) and it complained because the files didn't exist. Adding them let the test run without needing the files for some reason.

**Comment by bdbaddog** [comment]
Pushed updated testcase. Now uses directory fixture which has dirs and files to match sources/targets in test. 
Also added test for SOURCE.relpath.

File: SCons/Node/FS.py [hunk 2]
```diff
@@ -937,10 +947,11 @@ def _glob1(self, pattern, ondisk=True, source=False, strings=False):
 
 # Dict that provides a simple backward compatibility
 # layer for the Node attributes 'abspath', 'labspath',
-# 'path', 'tpath' and 'path_elements'.
+# 'relpath', 'path', 'tpath' and 'path_elements'.
 # @see Base.__getattr__ above
 node_bwcomp = {'abspath' : Base.get_abspath,
```
Comment (mwichmann) [inline SCons/Node/FS.py:952]: Strictly speaking, a new attribute doesn't have any "backward compatibility" concerns, but it probably doesn't hurt to have it here also.

File: SCons/__init__.py [hunk 3]
```diff
@@ -1,9 +1,9 @@
 __version__="4.1.1a"
 __copyright__="Copyright (c) 2001 - 2021 The SCons Foundation"
-__developer__="bdbaddog"
-__date__="2021-02-26 22:38:25"
-__buildsys__="ProDog2020"
-__revision__="3fa309f8fc5472fa067d7c04481fd0a5d3e86352"
-__build__="3fa309f8fc5472fa067d7c04481fd0a5d3e86352"
+__developer__="millsd"
```
Comment (mwichmann) [inline SCons/__init__.py:3]: Don't include this file in your commit. It's autogenerated on every full or doc build.

File: doc/design/engine.xml [hunk 4]
```diff
@@ -484,6 +484,17 @@
...
+    <term><literal>.relpath</literal></term>
+    <listitem>
+     <para>
+
+      The path to the file relative to the root SConstruct file's directory.
+
+     </para>
+    </listitem>
+   </varlistentry>
+
```
Comment (mwichmann) [inline doc/design/engine.xml:497]: We don't usually update these files, the design doc is woefully out of date.  That's more an FYI, if @bdbaddog doesn't object we can leave this here.

File: doc/man/scons.xml [hunk 5]
```diff
@@ -4925,6 +4925,13 @@ is not being used.</para>
   </listitem>
   </varlistentry>
 
+  <varlistentry>
+  <term><replaceable>n</replaceable>.<varname>relpath</varname></term>
```
Comment (mwichmann) [inline doc/man/scons.xml:4929]: I guess I'd have preferred the term `toprel` here - releative to the Top (aka SConstruct-containing) dir, as that term has started to creep in elsewhere (when describing `#path/to/something` form).  But if @bdbaddog as maintainer is okay with it, then I don't consider my opinion a blocker to the PR.

**Review by mwichmann — COMMENTED** [review]
I don't think any other place in the docs need touching, no.

File: SCons/Node/FS.py (hunk 2 above)
Comment (DeeeeLAN) [reply SCons/Node/FS.py:952]: Yeah I wasn’t certain about that part

File: doc/man/scons.xml (hunk 5 above)
Comment (DeeeeLAN) [reply doc/man/scons.xml:4929]: I have no objections to the rename, it is more self-documenting, and would leave the possibility of a relpath that isn’t necessarily relative to the top SConstruct.

Comment (bdbaddog) [reply doc/man/scons.xml:4929]: .relpath is better IMHO.. (maps to python API).

File: SCons/Node/FS.py (hunk 2 above)
Comment (bdbaddog) [reply SCons/Node/FS.py:952]: Good catch @mwichmann 

@DeeeeLan please remove the backward compat bit.

Comment (bdbaddog) [reply SCons/Node/FS.py:952]: fixed.

**Pull request merged**